cmd/snap-confine/ns-support.c

```diff
@@ -241,47 +242,107 @@ void sc_close_ns_group(struct sc_ns_group *group)
+				const char *base_snap_name,
+				const char *snap_name)
 {
 	// Open the mount namespace file.
 	char mnt_fname[PATH_MAX];
 	sc_must_snprintf(mnt_fname, sizeof mnt_fname, "%s%s", group->name,
 			 SC_NS_MNT_FILE);
 	int mnt_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
+
 	// NOTE: There is no O_EXCL here because the file can be around but
 	// doesn't have to be a mounted namespace.
 	//
 	// If the mounted namespace is discarded with
 	// sc_discard_preserved_ns_group() it will revert to a regular file.  If
 	// snap-confine is killed for whatever reason after the file is created but
 	// before the file is bind-mounted it will also be a regular file.
-	mnt_fd =
-	    openat(group->dir_fd, mnt_fname,
-		   O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
+	mnt_fd = openat(group->dir_fd, mnt_fname,
+			O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
 	if (mnt_fd < 0) {
 		die("cannot open mount namespace file for namespace group %s",
 		    group->name);
 	}
-	// Check if we got an nsfs-based file or a regular file. This can be
-	// reliably tested because nsfs has an unique filesystem type NSFS_MAGIC.
-	// On older kernels that don't support nsfs yet we can look for
-	// PROC_SUPER_MAGIC instead.
+	// Check if we got an nsfs-based or procfs file or a regular file. This can
+	// be reliably tested because nsfs has an unique filesystem type
+	// NSFS_MAGIC.  On older kernels that don't support nsfs yet we can look
+	// for PROC_SUPER_MAGIC instead. 
 	// We can just ensure that this is the case thanks to fstatfs.
-	struct statfs buf;
-	if (fstatfs(mnt_fd, &buf) < 0) {
-		die("cannot perform fstatfs() on an mount namespace file descriptor");
+	struct statfs ns_statfs_buf;
+	if (fstatfs(mnt_fd, &ns_statfs_buf) < 0) {
+		die("cannot perform fstatfs() on the mount namespace file descriptor");
+	}
+	// Stat the mount namespace as well, this is later used to check if the
+	// namespace is used by other processes if we are considering discarding a
+	// stale namespace.
+	struct stat ns_stat_buf;
+	if (fstat(mnt_fd, &ns_stat_buf) < 0) {
+		die("cannot perform fstat() on the mount namespace file descriptor");
 	}
 #ifndef NSFS_MAGIC
 // Account for kernel headers old enough to not know about NSFS_MAGIC.
 #define NSFS_MAGIC 0x6e736673
 #endif
-	if (buf.f_type == NSFS_MAGIC || buf.f_type == PROC_SUPER_MAGIC) {
+	if (ns_statfs_buf.f_type == NSFS_MAGIC
+	    || ns_statfs_buf.f_type == PROC_SUPER_MAGIC) {
+		char fname[PATH_MAX];
+		char base_snap_rev[PATH_MAX];
+
+		// Read the revision of the base snap.
+		sc_must_snprintf(fname, sizeof fname, "%s/%s/current",
+				 SNAP_MOUNT_DIR, base_snap_name);
+		if (readlink(fname, base_snap_rev, sizeof base_snap_rev) < 0) {
+			die("cannot read symlink %s", fname);
+		}
+		// Find the backing device of the base snap.
+		// TODO: add support for "try mode" base snaps that also need
+		// consideration of the mie->root component.
+		dev_t base_snap_dev = 0;
+		{
+			char base_squashfs_path[PATH_MAX];
```

Comment:
This looks like it might become a helper function?

Reply:
Yes, I was thinking the same thing. I just wanted to show the initial code to get quick evaluation before I started messing with it more closely.

Reply:
Done